### PR TITLE
Agency hasAsXXX using std::optional

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose

Old versions of `hasAsXXX` returned a `std::pair<XXX, bool>` where the first member is only safe to access if the second member is true. Most places did not check that. I replaced all of this with `std::optional<XXX>` and accessed the value using [`value()`](https://en.cppreference.com/w/cpp/utility/optional/value). `value()` throws an exception if the optional does not contain a value. Furthermore all the `hasAsXXX` methods used some `operator()` to obtain a node given a path. This method always threw an exception when the node did not exist. Since there are many keys that the supervision loop regularly accesses that intentionally do not exist, there are a lot of exceptions. Debugging with `catch throw` becomes impossible. This PR removed all those exceptions by returning a `std::nullopt` instead.

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [x] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [x] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

### Testing & Verification

*(Please pick either of the following options)*

- [x] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [x] This change is already covered by existing tests, such as all Supervision and Cluster Tests.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/15840/

File: arangod/Agency/Store.cpp [hunk 1]
```diff
@@ -418,7 +418,7 @@ check_ret_t Store::check(VPackSlice const& slice, CheckMode mode) const {
     // Check is guarded in ::apply
     bool found = _node.has(pv);
     if (found) {
-      node = &_node(pv);
+      node = &_node.get(pv)->get();
     }
```
Comment (goedderz) [inline arangod/Agency/Store.cpp:422]: Maybe rewrite to
```cpp
if (auto n = _node.get(pv); n) {
  node = &n->get();
}
```
?

File: arangod/Cluster/AgencyCache.cpp [hunk 2]
```diff
@@ -679,7 +679,7 @@ AgencyCache::change_set_t AgencyCache::changedSince(
   decltype(_planChanges) const& changes = (what == PLAN) ? _planChanges : _currentChanges;
 
   auto tmp = _readDB.nodePtr()->hasAsUInt(AgencyCommHelper::path(what) + "/" + VERSION);
-  uint64_t version = tmp.second ? tmp.first : 0;
+  uint64_t version = tmp ? *tmp : 0;
```
Comment (goedderz) [inline arangod/Cluster/AgencyCache.cpp:682]:
```suggestion
  uint64_t version = tmp.value_or(0);
```

**Review by goedderz — APPROVED** [review]

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/15991/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16000/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16020/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16041/

**Review by kvahed — APPROVED** [review]
Good Job

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16379/

**Pull request merged**